libsql/src/hrana/stream.rs

```diff
@@ -0,0 +1,372 @@
+            Err(HranaError::UnexpectedResponse(format!(
+                "Unexpected empty response from server: {:?}",
+                response.results
+            )))?;
+        }
+        if response.results.len() != N {
+            // One with actual results, one closing the stream
+            Err(HranaError::UnexpectedResponse(format!(
+                "Unexpected multiple responses from server: {:?}",
+                response.results
+            )))?;
+        }
+        let mut responses = std::array::from_fn(|_| StreamResponse::Close);
+        for i in 0..N {
+            match response.results.swap_remove(0) {
+                Response::Ok(StreamResponseOk { response }) => responses[i] = response,
+                Response::Error(e) => return Err(HranaError::StreamError(e)),
+            }
+        }
+        Ok(responses)
+    }
+
+    async fn finalize(&mut self, req: StreamRequest) -> Result<StreamResponse> {
+        let [resp, _] = self.send_requests([req, StreamRequest::Close]).await?;
+        self.done();
+        Ok(resp)
+    }
+
+    fn done(&mut self) {
+        if let Some(baton) = &self.baton {
+            tracing::trace!("closing client stream (baton: `{}`)", baton);
+        }
+        self.baton = None;
+        self.sql_id_generator = 0;
+        self.status = StreamStatus::Closed;
+    }
+
+    async fn close(&mut self) -> Result<()> {
+        if self.status == StreamStatus::Closed {
+            return Ok(());
+        }
+        self.send(StreamRequest::Close).await?;
+        self.done();
+        Ok(())
+    }
+
+    fn next_sql_id(&mut self) -> SqlId {
+        self.sql_id_generator = self.sql_id_generator.wrapping_add(1);
+        self.sql_id_generator
+    }
+}
+
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub enum StreamStatus {
+    Open,
+    Closed,
+}
+
+#[derive(Debug, Clone)]
+pub struct StoredSql<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    stream: HttpStream<T>,
+    sql_id: SqlId,
+}
+
+impl<T> StoredSql<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    fn new(stream: HttpStream<T>, sql_id: SqlId) -> Self {
+        StoredSql { stream, sql_id }
+    }
+
+    pub async fn describe(&self) -> Result<DescribeResult> {
+        self.stream.describe(self).await
+    }
+
+    pub async fn execute(&self) -> Result<()> {
+        self.stream.sequence(self).await
+    }
+
+    pub async fn close(self) -> Result<()> {
+        self.stream.close_sql(self.sql_id).await
+    }
+}
+
+pub trait SqlDescriptor {
+    fn sql_description(&self) -> SqlDescription;
+}
+
+impl SqlDescriptor for String {
+    fn sql_description(&self) -> SqlDescription {
+        SqlDescription::Sql(self.clone())
+    }
+}
+
+impl<T> SqlDescriptor for StoredSql<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    fn sql_description(&self) -> SqlDescription {
+        SqlDescription::SqlId(self.sql_id)
+    }
+}
+
+#[derive(Debug)]
+pub enum SqlDescription {
+    Sql(String),
+    SqlId(SqlId),
```

Comment:
What about String and Remote? Instead of these two names that seem quite similar?

Reply:
I named them this way, since they map directly onto Hrana protocol fields.